x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtRealm.java

```diff
@@ -403,7 +403,7 @@ private void processValidatedJwt(
                     () -> format("Realm [%s] roles resolved [%s] for principal=[%s].", name(), join(",", user.roles()), principal)
                 );
                 if (isCacheEnabled()) {
-                    try (ReleasableLock ignored = jwtCacheHelper.acquireUpdateLock()) {
+                    try (ReleasableLock ignored = jwtCacheHelper.acquireWriteLock()) {
```

Comment:
Per my other comment, I think this is wrong. The `Cache.put` should use the reentrant lock not the exclusive lock. Only the `invalidateAll` needs exclusive access.

Reply:
Thanks Tim. This makese sense. Sorry I didn't think it carefully enough. @jakelandis I will go ahead and make changes to this PR. I hope you don't mind. I'll iterate it with Tim to hopefully get this PR merged during my day time.

Reply:
thanks for catching this.